jgrapht-core/src/main/java/org/jgrapht/alg/tour/WarnsdorffRuleKnightTourHeuristic.java

```diff
@@ -0,0 +1,810 @@
+     */
+
+    public KnightTour getTour(TourType type, boolean structured, int shiftX, int shiftY) {
+
+        if (shiftX < 0 || shiftY < 0) {
+            throw new UnsupportedOperationException("You have set incorrect shift!");
+        }
+
+        if (!checkExistence(type)) {
+            throw new UnsupportedOperationException("No solution exist for such configuration!");
+        }
+
+        KnightTour tour = new KnightTour();
+        Random rand = new Random();
+        int startX, startY;
+        Pair<Integer, Integer> currentCell = new Pair<>(-1, -1);
+        int visited;
+        int run = 0;
+
+        boolean[][] wasStartingVertex = new boolean[n][m];
+
+        boolean found = false;
+        while (!found) {
+            visited = 0;
+
+            for (int i = 0; i < n; i++) {
+                for (int j = 0; j < m; j++) {
+                    chessBoard[i][j] = false;
+                }
+            }
+
+            tour.getList().clear();
+
+            startX = rand.nextInt(n);
+            startY = rand.nextInt(m);
+
+            currentCell.setFirst(startX);
+            currentCell.setSecond(startY);
+
+            while (wasStartingVertex[startX][startY]) {
+                startX = rand.nextInt(n);
+                startY = rand.nextInt(m);
+                currentCell.setFirst(startX);
+                currentCell.setSecond(startY);
+            }
+
+            wasStartingVertex[startX][startY] = true;
+            run++;
+
+            while (visited < n * m) {
+                chessBoard[currentCell.getFirst()][currentCell.getSecond()] = true;
+                tour.getList().add(currentCell);
+
+                /*
+                 If we have added the structured cell then update pointer on that cell in the structured array.
+                 */
+
+                if (structured) {
+                    int val = updateStructuredPosition(currentCell);
+                    if (val != -1) {
+                        tour.getStructured().set(val, tour.getList().getTail());
+                    }
+                }
+
+                visited++;
+                currentCell = getMoveWarnsdorff(currentCell);
+                if (currentCell.getFirst() == -1) {
+                    break;
+                }
+            }
+
+            Pair<Integer, Integer> endCell = tour.getList().getTail().getValue();
+            if (visited == n * m && checkType(startX, startY, endCell.getFirst(), endCell.getSecond(), type)) {
+                HashSet<Pair<Pair<Integer, Integer>, Pair<Integer, Integer>>> moves = getMoves(tour.getList());
+                if (checkStructured(moves, structured)) {
+                    found = true;
+                }
+            }
+
+            /*
+             Try again if there is no unused start cells are left.
+             */
+
+            if (run == (n * m) && ! found) {
+                throw new UnsupportedOperationException("The Warnsdorff's rule heuristic has failed to find a tour!");
```

Comment:
Probably return a 'null' tour and not an exception. It is not really something that should crash the user's application.